File: 17_Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, d: str) -> List[str]:
        mapping = {'2': ['a', 'b', 'c'],
                 '3': ['d', 'e', 'f'],
                 '4': ['g', 'h', 'i'],
                 '5': ['j', 'k', 'l'],
                 '6': ['m', 'n', 'o'],
                 '7': ['p', 'q', 'r', 's'],
                 '8': ['t', 'u', 'v'],
                 '9': ['w', 'x', 'y', 'z']}
        results=[[]]
                
        for digit in d:
            temp=[]
           
            for result in results:
                
                for letter in mapping[digit]:
                    temp.append(result + [letter])
                    
                    
            results = temp
            
            
        return ["".join(result) for result in results]
                
                
            
            
            
            
            
        
                
                
                
                
        

File: 17_Letter_Combinations_of_a_Phone_Number.cpp
class Solution {

    unordered_map<char,string> mp;
	vector<string> result;
public:

	void dfs(string& digits, int ind, string str){
		if(ind==digits.size()){
			result.push_back(str);
            return;
		}
		for(auto& a:m[digits[ind]]){
			if(a=='1') continue;
			dfs(digits,index+1,path+a);
		}
	}
	vector<string> letterCombinations(string digits) {
		if(digits.size()==0) return ans;
		mp['2']="abc";
		mp['3']="def";
		mp['4']="ghi";
		mp['5']="jkl";
		mp['6']="mno";
		mp['7']="pqrs";
		mp['8']="tuv";
		mp['9']="wxyz";
		dfs(digits,0,"");
		return result;
	}
};
